File: extension/case-engine.js
// Shift+F3 case engine — content-script copy (no ES module syntax).
// Mirror of /web/case-engine.js so changes there should be reflected here.
(function (root) {
  const CYCLE = ['invert', 'lower', 'upper', 'title', 'sentence'];

  const SMALL_WORDS = new Set([
    'a', 'an', 'and', 'as', 'at', 'but', 'by', 'for', 'if', 'in',
    'nor', 'of', 'on', 'or', 'so', 'the', 'to', 'up', 'yet', 'via'
  ]);
  const KEEP_VERBATIM = ['I', 'AHK', 'AutoHotkey'];

  function invert(s) {
    let out = '';
    for (const ch of s) {
      const lo = ch.toLowerCase();
      const up = ch.toUpperCase();
      if (ch === lo && ch !== up) out += up;
      else if (ch === up && ch !== lo) out += lo;
      else out += ch;
    }
    return out;
  }
  function lower(s) { return s.toLowerCase(); }
  function upper(s) { return s.toUpperCase(); }
  function title(s) {
    return s.replace(/(\S+)/g, (word, _m, offset, full) => {
      const lc = word.toLowerCase();
      const stripped = lc.replace(/[^a-z']/gi, '');
      const atStart = offset === 0 || /[\.\!\?]\s*$/.test(full.slice(0, offset));
      if (!atStart && SMALL_WORDS.has(stripped)) return lc;
      return lc.replace(/[a-zà-ÿ]/, c => c.toUpperCase());
    }).replace(/\b([a-z]+)\b/gi, (m) => {
      for (const k of KEEP_VERBATIM) {
        if (m.toLowerCase() === k.toLowerCase()) return k;
      }
      return m;
    });
  }
  function sentence(s) {
    return s.toLowerCase().replace(
      /(^|[\.\!\?]\s+|[\r\n]+)(\S)/g,
      (_m, lead, ch) => lead + ch.toUpperCase()
    );
  }
  const TRANSFORMS = { invert, lower, upper, title, sentence };
  function apply(name, t) { return TRANSFORMS[name](t); }
  function nextStateFor(t, state) {
    const name = CYCLE[state];
    return { text: apply(name, t), name, state: (state + 1) % CYCLE.length };
  }

  root.CaseEngine = { CYCLE, apply, invert, lower, upper, title, sentence, nextStateFor };
})(typeof self !== 'undefined' ? self : this);

File: desktop/case_engine.py
"""Pure-Python implementation of the Shift+F3 case cycle.

Mirrors the JS engine in ``web/case-engine.js`` and the AHK behavior so
the desktop daemon, the browser extension, and the web app all produce
the same output for a given input.
"""

from __future__ import annotations

import re

CYCLE = ("invert", "lower", "upper", "title", "sentence")

_SMALL_WORDS = {
    "a", "an", "and", "as", "at", "but", "by", "for", "if", "in",
    "nor", "of", "on", "or", "so", "the", "to", "up", "yet", "via",
}
_KEEP_VERBATIM = {"i": "I", "ahk": "AHK", "autohotkey": "AutoHotkey"}


def invert(s: str) -> str:
    return "".join(
        c.upper() if c.islower() else c.lower() if c.isupper() else c
        for c in s
    )


def lower(s: str) -> str:
    return s.lower()


def upper(s: str) -> str:
    return s.upper()


_WORD_RE = re.compile(r"\S+")
_LETTER_RE = re.compile(r"[a-zà-ÿ]", re.IGNORECASE)


def title(s: str) -> str:
    def repl(match: re.Match[str]) -> str:
        word = match.group(0)
        offset = match.start()
        lc = word.lower()
        stripped = re.sub(r"[^a-z']", "", lc, flags=re.IGNORECASE)
        before = s[:offset]
        at_start = offset == 0 or bool(re.search(r"[\.\!\?]\s*$", before))
        if not at_start and stripped in _SMALL_WORDS:
            piece = lc
        else:
            piece = _LETTER_RE.sub(lambda m: m.group(0).upper(), lc, count=1)
        if stripped in _KEEP_VERBATIM:
            piece = re.sub(
                rf"\b{re.escape(stripped)}\b",
                _KEEP_VERBATIM[stripped],
                piece,
                flags=re.IGNORECASE,
            )
        return piece

    return _WORD_RE.sub(repl, s)


_SENTENCE_RE = re.compile(r"(^|[\.\!\?]\s+|[\r\n]+)(\S)")


def sentence(s: str) -> str:
    lc = s.lower()
    return _SENTENCE_RE.sub(lambda m: m.group(1) + m.group(2).upper(), lc)


_TRANSFORMS = {
    "invert": invert,
    "lower": lower,
    "upper": upper,
    "title": title,
    "sentence": sentence,
}


def apply(name: str, text: str) -> str:
    return _TRANSFORMS[name](text)


def next_state(text: str, state: int) -> tuple[str, str, int]:
    """Apply ``CYCLE[state]`` to ``text`` and return ``(text, name, next)``."""
    name = CYCLE[state]
    return apply(name, text), name, (state + 1) % len(CYCLE)
